Orange/widgets/visualize/utils/widget.py

```diff
@@ -0,0 +1,623 @@
+
+    def _add_graph(self):
+        box = gui.vBox(self.mainArea, True, margin=0)
+        self.graph = self.GRAPH_CLASS(self, box)
+        box.layout().addWidget(self.graph.plot_widget)
+
+    def _add_top_controls(self):
+        self.graph.gui.point_properties_box(self.controlArea)
+
+    def _add_middle_controls(self):
+        self._effects_box = self.graph.gui.effects_box(self.controlArea)
+        self._plot_box = self.graph.gui.plot_properties_box(self.controlArea)
+
+    def _add_bottom_controls(self):
+        self.controlArea.layout().addStretch(100)
+        self.graph.box_zoom_select(self.controlArea)
+        gui.auto_commit(self.controlArea, self, "auto_commit",
+                        "Send Selection", "Send Automatically")
+
+    # Input
+    @Inputs.data
+    @check_sql_input
+    def set_data(self, data):
+        self.closeContext()
+        self.clear()
+        self.data = data
+        self.check_data()
+        self.init_attr_values()
+        self.openContext(self.data)
+        self.cb_class_density.setEnabled(self.can_draw_density())
+
+    def check_data(self):
+        """Handle error messages and setting self.valid_data property.
+        Should be overridden most of the times when subclassed."""
+        self.clear_messages()
+        if self.data is not None:
+            if self.data.is_sparse():
+                self.Error.sparse_data()
+                self.data = None
+        self.valid_data = np.any(np.isfinite(self.data.X), 1) \
+            if self.data is not None else None
+        if self.data is not None and not len(self.data[self.valid_data]):
+            self.data = None
+            self.Warning.missing_coords()
+
+    @Inputs.data_subset
+    @check_sql_input
+    def set_subset_data(self, subset):
+        self.subset_data = subset
+        self.subset_indices = {e.id for e in subset} \
+            if subset is not None else {}
+        self.controls.graph.alpha_value.setEnabled(subset is None)
+
+    def handleNewSignals(self):
+        if self.data is not None:
+            self.init_embedding_coords()
+            self.setup_plot()
+            self.apply_selection()
+        self.commit()
+
+    def get_subset_mask(self):
+        if self.subset_indices:
+            return np.array([ex.id in self.subset_indices
+                             for ex in self.data[self.valid_data]])
+        return None
+
+    # Plot
+    def init_embedding_coords(self):
+        """Setting self._embedding_coords.
+        Should be overridden when subclassed."""
+        x_data = self.data.X
+        x_data[x_data == np.inf] = np.nan
+        x_data = np.nanmean(x_data[self.valid_data], 1)
+        y_data = np.ones(len(x_data))
+        self._embedding_coords = np.vstack((x_data, y_data)).T
+
+    def get_coordinates_data(self):
+        return (self._embedding_coords[:, 0], self._embedding_coords[:, 1]) \
+            if self._embedding_coords is not None else (None, None)
```

Comment:
Maybe `return self._embedding_coords.T if self._embedding_coords is not None else (None, None)`? It makes it more obvious that it returns the complete `_embedding_coords` and not just the first two columns.